plugins/uefi-dbx/fu-uefi-dbx-device.c

```diff
@@ -0,0 +1,153 @@
+	buf = g_bytes_get_data (fw, &bufsz);
+	if (!fu_efivar_set_data (FU_EFIVAR_GUID_SECURITY_DATABASE,
+				 "dbx", buf, bufsz,
+				 FU_EFIVAR_ATTR_APPEND_WRITE |
+				 FU_EFIVAR_ATTR_TIME_BASED_AUTHENTICATED_WRITE_ACCESS |
+				 FU_EFIVAR_ATTR_RUNTIME_ACCESS |
+				 FU_EFIVAR_ATTR_BOOTSERVICE_ACCESS |
+				 FU_EFIVAR_ATTR_NON_VOLATILE,
+				 error)) {
+		return FALSE;
+	}
+
+	/* success! */
+	return TRUE;
+}
+
+static gboolean
+fu_uefi_dbx_device_set_version_number (FuDevice *device, GError **error)
+{
+	gsize bufsz = 0;
+	guint csum_cnt = 0;
+	g_autofree gchar *version = NULL;
+	g_autofree guint8 *buf = NULL;
+	g_autoptr(GPtrArray) dbx = NULL;
+	const gchar *ignored_guids[] = {
+		FU_EFI_SIGNATURE_GUID_OVMF,
+		NULL };
+
+	/* use the number of checksums in the dbx as a version number, ignoring
+	 * some owners that do not make sense */
+	if (!fu_efivar_get_data (FU_EFIVAR_GUID_SECURITY_DATABASE, "dbx",
+				 &buf, &bufsz, NULL, error))
+		return FALSE;
+	dbx = fu_efi_signature_parser_new (buf, bufsz,
+					   FU_EFI_SIGNATURE_PARSER_FLAGS_NONE,
+					   error);
+	if (dbx == NULL)
+		return FALSE;
+	for (guint j = 0; j < dbx->len; j++) {
+		FuEfiSignatureList *siglist = g_ptr_array_index (dbx, j);
+		GPtrArray *items = fu_efi_signature_list_get_all (siglist);
+		for (guint i = 0; i < items->len; i++) {
+			FuEfiSignature *sig = g_ptr_array_index (items, i);
+			if (fu_efi_signature_get_kind (sig) != FU_EFI_SIGNATURE_KIND_SHA256)
+				continue;
+			if (g_strv_contains (ignored_guids, fu_efi_signature_get_owner (sig)))
+				continue;
+			csum_cnt++;
+		}
+	}
+	version = g_strdup_printf ("%u", csum_cnt);
+	fu_device_set_version (device, version);
+	fu_device_set_version_lowest (device, version);
+	return TRUE;
+}
+
+static gboolean
+fu_uefi_dbx_device_probe (FuDevice *device, GError **error)
+{
+	gsize bufsz = 0;
+	g_autofree gchar *devid = NULL;
+	g_autofree guint8 *buf = NULL;
+	g_autoptr(GPtrArray) kek = NULL;
+
+	/* use each of the certificates in the KEK to generate the GUIDs */
+	if (!fu_efivar_get_data (FU_EFIVAR_GUID_EFI_GLOBAL, "KEK",
+				 &buf, &bufsz, NULL, error))
+		return FALSE;
+	kek = fu_efi_signature_parser_new (buf, bufsz,
+					   FU_EFI_SIGNATURE_PARSER_FLAGS_NONE,
+					   error);
+	if (kek == NULL)
+		return FALSE;
+	for (guint i = 0; i < kek->len; i++) {
+		FuEfiSignatureList *siglist = g_ptr_array_index (kek, i);
+		GPtrArray *sigs = fu_efi_signature_list_get_all (siglist);
+		for (guint j = 0; j < sigs->len; j++) {
+			FuEfiSignature *sig = g_ptr_array_index (sigs, j);
+			fu_device_add_instance_id (device, fu_efi_signature_get_checksum (sig));
+		}
+	}
+	return fu_uefi_dbx_device_set_version_number (device, error);
+}
+
+static void
+fu_uefi_dbx_device_init (FuUefiDbxDevice *self)
+{
+	fu_device_set_physical_id (FU_DEVICE (self), "dbx");
+	fu_device_set_name (FU_DEVICE (self), "UEFI dbx");
+	fu_device_set_summary (FU_DEVICE (self), "UEFI dbx Virtual Device");
+	fu_device_set_vendor (FU_DEVICE (self), "Microsoft");
```

Comment:
I don't think this is reasonable to set here - should be looking at what's in KEK and export it based on that (it would actually be Microsoft Corporation for most systems)

Reply:
> should be looking at what's in KEK

I'm not sure I actually want to be *parsing* the X509 certificate, although I suppose we might have to start doing that in the future for other things. If we did the vendor would be `Microsoft Corporation KEK CA 2011` which probably isn't what the user would expect. Maybe it's best we don't set a vendor at all. Keep reading...

The more important one to get right is the `vendor_id` and with a brain turned on hardcoding `UEFI:Microsoft` isn't smart either. I briefly thought we want to use the KEK checksum again, although then the question becomes _which_ certificate in the KEK should be used. I think hardcoding a heuristic would be pretty nasty given the OEM usually puts a cert there too.

Given I think the way this should work is by having a super-locked down LVFS vendor group (with just the DBX updates there, only pushable by LVFS admins...) so maybe having a vendor_id of `"LVFS"` or `"unspecified"` isn't so bad.

Reply:
>I'm not sure I actually want to be parsing the X509 certificate, although I suppose we might have to start doing that in the future for other things. If we did the vendor would be Microsoft Corporation KEK CA 2011 which probably isn't what the user would expect. Maybe it's best we don't set a vendor at all. Keep reading...

I think it would be the "O" field not the "CN" field.
`Subject: C=US, ST=Washington, L=Redmond, O=Microsoft Corporation, CN=Microsoft Corporation KEK CA 2011`

>Given I think the way this should work is by having a super-locked down LVFS vendor group (with just the DBX updates there, only pushable by LVFS admins...) so maybe having a vendor_id of "LVFS" or "unspecified" isn't so bad.

I think until there is a third party we know of actually _uploading_ a dbx to LVFS, how about "Linux Foundation"?

Reply:
> I think it would be the "O" field

Of which certificate tho?

> how about "Linux Foundation"?

So `UEFI:Linux Foundation` with a _space_? I guess that's okay, it just looked weird.

Reply:
>Of which certificate tho?

I pulled that from the one in key 2 in my KEK

>So UEFI:Linux Foundation with a space? I guess that's okay, it just looked weird.

Dell's is `(DMI: Dell Inc.)`

Reply:
> I pulled that from the one in key 2 in my KEK

But how did you decide to use the 2nd key and not the 1st?